mantidproject/mantid
# A utility WorkflowAlgorithmRunner to manage data reduction workflows

The purpose of this algorithm is to do dependency resolution, data flow control, and execution management for another algorithm. It is mainly intended to be used later with certain data reduction workflow algorithms at ILL; however, it can run any type of algorithm if needed.

Implements #17480 . For further documentation, see [this](https://github.com/mantidproject/documents/blob/master/Project-Management/ILL/WorkspaceAlgorithmRunner.md).

**To test:**
The algorithm needs a bit of setup. I'll use the Python example from the documentation here; it should provide a nice starting point for testing.

``` python
# This is our initial workspace. We want to scale it first by 100
# and then by 0.42
CreateSingleValuedWorkspace(OutputWorkspace='ws2', DataValue=1.0)

# Setup the runs for the Scale algorithm
setupTable = WorkspaceFactoryImpl.Instance().createTable()
setupTable.addColumn('str', 'Run name')
# The rest of the columns can be in arbitrary order
setupTable.addColumn('str', 'InputWorkspace')
setupTable.addColumn('double', 'Factor')
setupTable.addColumn('str', 'OutputWorkspace')
row = {
    'Run name': '1st run',
    'InputWorkspace': '2nd run', # Takes the input of '2nd run'.
    'Factor': 0.42,
    'OutputWorkspace': '"ws1"' # Forced output. Otherwise Scale fails to execute.
}
setupTable.addRow(row)
row = {
    'Run name': '2nd run',
    'InputWorkspace': "'ws2'", # Hard-coded input
    'Factor': 100,
    'OutputWorkspace': 'scaled_ws2'
}
setupTable.addRow(row)
AnalysisDataServiceImpl.Instance().addOrReplace('setupTable', setupTable)

# Map the Scale algorithm's OutputWorkspace to InputWorkspace
ioMap = WorkspaceFactoryImpl.Instance().createTable()
ioMap.addColumn('str', 'InputWorkspace')
ioMap.addRow({'InputWorkspace': 'OutputWorkspace'})
AnalysisDataServiceImpl.Instance().addOrReplace('ioMapTable', ioMap)

# Execute the algorithm
WorkflowAlgorithmRunner('Scale', SetupTable=setupTable, InputOutputMap=ioMap)

# Print some results
print('Original input value: {0}'.format(mtd['ws2'].dataY(0)[0]))
print('After scaling by 100: {0}'.format(mtd['scaled_ws2'].dataY(0)[0]))
print('After further scaling by 0.42: {0}'.format(mtd['ws1'].dataY(0)[0]))
```

The above script should print results 1, 100, and 42.

_Release notes were updated._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
